### PR TITLE
planner: complete the memo/group-expression/group logic and add test for derive stats

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/51664

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/cascades/memo/memo.go [hunk 1]
```diff
@@ -19,17 +19,13 @@ import (
...
 )
 
 // Memo is the main structure of the memo package.
 type Memo struct {
-	// ctx is the context of the memo.
-	sCtx sessionctx.Context
-
 	// groupIDGen is the incremental group id for internal usage.
-	groupIDGen GroupIDGenerator
+	groupIDGen *GroupIDGenerator
```
Comment (AilinKid) [inline pkg/planner/cascades/memo/memo.go:28]: according the detail design, memo/rule should be not aware of session context, but scheduler and task did

File: pkg/planner/core/casetest/cascades/testdata/cascades_suite_out.json [hunk 2]
```diff
@@ -0,0 +1,107 @@
+[
+  {
+    "Name": "TestDeriveStats",
+    "Cases": [
+      {
+        "SQL": "select count(1) from t1 group by a, b",
+        "Str": "GID:1, logic prop:{stats:{count 5, ColNDVs map[1:2 2:2]}, schema:{Column: [test.t1.a,test.t1.b] Key: [] Unique key: []}}\nGID:2, logic prop:{stats:{count 4, ColNDVs map[4:4]}, schema:{Column: [Column#4] Key: [] Unique key: []}}\nGID:3, logic prop:{stats:{count 4, ColNDVs map[4:4]}, schema:{Column: [Column#4] Key: [] Unique key: []}}"
```
Comment (winoros) [inline pkg/planner/core/casetest/cascades/testdata/cascades_suite_out.json:7]: The `\n` seems to be parsed in the wrong way?

Comment (AilinKid) [reply pkg/planner/core/casetest/cascades/testdata/cascades_suite_out.json:7]: json string seems couldn't switch line, i change it to string slice, which will be much more clear
image

File: pkg/planner/cascades/memo/group_expr.go [hunk 3]
```diff
@@ -122,3 +124,39 @@ func (e *GroupExpression) Init(h base2.Hasher) {
...
+// has already gotten its logical prop.
+func (e *GroupExpression) DeriveLogicalProp() (err error) {
+	if e.GetGroup().GetLogicalProperty() != nil {
+		return nil
+	}
+	childStats := make([]*property.StatsInfo, 0, len(e.Inputs))
+	childSchema := make([]*expression.Schema, 0, len(e.Inputs))
+	for _, childG := range e.Inputs {
+		childGProp := childG.GetLogicalProperty()
+		intest.Assert(childGProp != nil)
```
Comment (hawkingrei) [inline pkg/planner/cascades/memo/group_expr.go:140]: Can we add this ```assert``` into ```GetLogicalProperty```？

Comment (AilinKid) [reply pkg/planner/cascades/memo/group_expr.go]: make sense

File: pkg/planner/cascades/memo/group.go [hunk 4]
```diff
@@ -127,9 +128,47 @@ func (g *Group) GetFirstElem(operand pattern.Operand) *list.Element {
...
+
+// SetExplored set the group as tagged as explored.
+func (g *Group) SetExplored() {
+	g.explored = true
+}
+
 // String implements fmt.Stringer interface.
 func (g *Group) String(w util.StrBufferWriter) {
-	w.WriteString(fmt.Sprintf("inputs:%s", strconv.Itoa(int(g.groupID))))
+	w.WriteString(fmt.Sprintf("GID:%s", strconv.Itoa(int(g.groupID))))
```
Comment (hawkingrei) [inline pkg/planner/cascades/memo/group.go:159]:
```suggestion
	w.WriteString(fmt.Sprintf("GID:%d", int(g.groupID)))
```

**Review by hawkingrei — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**